cpp/src/arrow/result.h

```diff
@@ -313,6 +317,29 @@ class Result : public util::EqualityComparable<Result<T>> {
     return Status::OK();
   }
 
+  T ValueOr(T alternative) && {
+    if (!ok()) {
+      return alternative;
+    }
+    return std::move(arrow::util::get<T>(variant_));
+  }
+
+  template <typename M>
+  Result<typename std::result_of<M && (T)>::type::ValueType> Map(M&& m) && {
```

Comment:
Hmm... Isn't `Result<typename std::result_of<M && (T)>::type::ValueType>` the same thing as `typename std::result_of<M && (T)>::type`?

Reply:
It is, this is a weaksauce guarantee that M returns a Result